### ISSUE TITLE
[Question] code question in `quip.quantize_linear`

### ISSUE THREAD

**Issue by jhss** [Cornell-RelaxML/quip-sharp#75]
Hello.

I'm trying to understand your code related to the QuIP paper, specifically how the Hessian and weights are processed.

In the QuIP paper (not QuIP#), Hessian and weights undergo with the following algorithm:

![8](https://github.com/user-attachments/assets/2448e6c5-3cd1-43ee-9bfe-6a6435ec4fd3)

In your code (`quip-shapr/lib/algo/quip/quantize_linear`), the Hessian and weights are processed as follows:
```python3
H_data = torch.load(hessian_path, map_location=torch.device('cpu'))
H = utils.flat_to_sym(H_data['flatH'], H_data['n']) # 
mu = H_data['mu']
H.add_(mu[None, :] * mu[:, None]) # Pre-Processing Step 2 (?)
n = H_data['n']
W = torch.vstack([
    weights[i].to(dtype_) / scales[i] for i in range(len(weights))
]).to(dtype_) # ?
H = utils.regularize_H(H, n, args.sigma_reg) # ?
hatW, attr = quantize(H, W, args.lora_rank, cb, args, device) # call incoherence_preprocess
```
In the incoherence_preprocess function, it seems that the code corresponds to Steps 3 and 4 of the QuIP Incoherence Pre-Processing Algorithm:

```python3
def incoherence_preprocess(H, W, args):
    ...
    # QuIP - InCoherence Pre-Processing Step 3~4
    if args.rescale_WH:
        Hr = H / H.abs().max() # ?
        diagH = torch.diag(Hr)
        diagW2 = torch.diag(W.T @ W)
        diagH = torch.clamp(diagH, min=1e-8)
        diagW2 = torch.clamp(diagW2, min=1e-8)
        scaleWH = (diagH / diagW2).sqrt().sqrt().to(torch.float32)
        scaleWH = scaleWH.clamp(min=1e-8)
        Wr = Wr * scaleWH[None, :]
        Hr = Hr / scaleWH[None, :]
        Hr = Hr / scaleWH[:, None]
        scaleWH = scaleWH.cpu()
   ...
```

In the code above, I'm having difficulty understanding certain parts of the code:

#### (1) `H.add_(mu[None, :] * mu[:, None])`

I believe this line is related with Step 2 in QuIP Incoherence Pre-Processing Algorithm.
In the paper, you add `alpha * mean(diag(H))I`, but in the code, your adding $$mu \cdot mu^T$$.
Could you explain why do you add $$mu \cdot mu^T$$ instead of alpha * mean(diag(H))I?

#### (2)  `W = torch.vstack([weights[i].to(dtype_) / scales[i] for i in range(len(weights))]).to(dtype_)`

I think this line corresponds to Step 4 of the Algorithm, where the weights are rescaled by $$\tilde{D}$$.
In the code, the weights are divided by scales, which seems to be their L2 norms.
Why are you dividing by the L2 norms before applying the rescaling with $$\tilde{D}$$?

#### (3) `H = utils.regularize_H(H, n, args.sigma_reg)` and `Hr = H / H.abs().max()`

In the QuIP paper, before applying the pre-processing step ($$VHV^T$$), the Hessian is only adjusted by adding `alpha * mean(daig(H))I`. In your code, it appears that additional steps are taken: regularization (`regularize_H`) and normalization by dividing `H` by its absolute maximum. Could you clarify why these additional steps (regularization and normalization) are included?

I might be missing some details from the paper, but I would appreciate it if you could help me understand these points.

Thank you!

**Comment by tsengalb99** [comment]
1) We subtract mu * mu.T when we save the hessians so this reverses that (see [here](https://github.com/Cornell-RelaxML/quip-sharp/blob/d12f479daf3a63ed2841dd00d4f4d222cafa739b/quantize_llama/hessian_offline_llama.py#L112)). I don't remember why we did this and its not necessary. The QTIP code gets rid of this.
2) This is for when we fuse different layers together that share the same input (e.g. q, k, and v). This scales all the layers so that they have "equal contribution" in the incoherence processing step. If you're not fusing layers then this step isn't necessary. IIRC not doing this step also isn't catastrophic, but it may hurt quality a bit.
3) regularize_H just adds a small diagonal bias to make the matrix PD. Even though H is PD in theory, the empirically collected H may not be due to numerical errors. Likewise, we divide H by the largest entry to mitigate numerical overflow. If you use fp64 during the quantization pipeline (`--fp64` flag) you might be able to skip these steps or reduce the strength of the regularization.
